src/Tuner.cpp

```diff
@@ -0,0 +1,424 @@
+#define TUNER_FILE_LOCAL "leelaz_tuners"
+#define MAX_ERROR 1e-4
+
+using namespace Utils;
+
+static void sgemmBatched_ref(const std::vector<float>& a, const std::vector<float>& b,
+        std::vector<float>& c, const int m, const int n, const int k,
+        const int batch_size) {
+
+    for (auto batch = 0; batch < batch_size; batch++) {
+
+        auto offset_u = batch*m*k;
+        auto offset_v = batch*n*k;
+        auto offset_m = batch*m*n;
+
+        cblas_sgemm(CblasRowMajor, CblasTrans, CblasNoTrans,
+                    m, n, k,
+                    1.0f,
+                    &a[offset_u], m,
+                    &b[offset_v], n,
+                    0.0f,
+                    &c[offset_m], n);
+    }
+}
+
+
+bool IsMultiple(const size_t a, const size_t b) {
+    return ((a/b)*b == a);
+};
+
+bool Tuner::valid_config_sgemm(Parameters p) {
+
+    if (!IsMultiple(p["MWG"], p["MDIMC"]*p["VWM"])) {
+        return false;
+    }
+    if (!IsMultiple(p["NWG"], p["NDIMC"]*p["VWN"])) {
+        return false;
+    }
+    if (!IsMultiple(p["MWG"], p["MDIMA"]*p["VWM"])) {
+        return false;
+    }
+    if (!IsMultiple(p["NWG"], p["NDIMB"]*p["VWN"])) {
+        return false;
+    }
+    if (!IsMultiple(p["KWG"], p["MDIMC"]*p["NDIMC"]/p["MDIMA"])) {
+        return false;
+    }
+    if (!IsMultiple(p["KWG"], p["MDIMC"]*p["NDIMC"]/p["NDIMB"])) {
+        return false;
+    }
+    if (p["MDIMC"] != p["MDIMA"]) {
+        return false;
+    }
+    if (p["NDIMC"] != p["NDIMB"]) {
+        return false;
+    }
+    if (p["SA"] != p["SB"]) {
+        return false;
+    }
+    return true;
+}
+
+
+Parameters Tuner::get_parameters_by_int(const std::vector<Configurations> opts, const int n) {
+
+        Parameters param;
+        std::vector<size_t> choices(opts.size());
+
+        auto cfgs = 1;
+        for (auto c = size_t{0}; c < opts.size(); c++) {
+            choices[c] = opts[c].second.size();
+            cfgs *= choices[c];
+        }
+        int j = n;
+
+        for (auto c = size_t{0}; c < opts.size(); c++) {
+            auto o = opts[c];
+            auto s = o.first;
+            auto v = o.second[j % choices[c]];
+            j /= choices[c];
+            param[s] = v;
+        }
+
+        return param;
+}
+
+std::string Tuner::parameters_to_string(const Parameters p) {
+    std::string s;
+    for (auto const& x : p) {
+        s += " -D" + x.first + "=" + std::to_string(x.second);
+    }
+    return s;
+}
+
+size_t next_power_of_two(const size_t x) {
+    return 2 << (size_t)(std::ceil(std::log2(x)) - 1);
+}
+
+static void sgemm_generate_data(std::vector<float> &x, const int m, const int n,
+        const int batch_size, const int m_ceil, const int n_ceil) {
+    for (auto batch = 0; batch < batch_size; batch++) {
+        for (auto i = 0; i < n_ceil; i++) {
+            if (i < n) {
+                for (auto j = 0; j < m; j++) {
+                    x[batch*n_ceil*m_ceil + i*m_ceil + j] = 0.01f*(((i ^ j) + batch - 50) % 100);
```

Comment:
Won't i ^ j overflow? Why not random floats?

Reply:
Isn't ^ xor operation for ints? I can't use the builtin random generator here, because it would mess the reproducibility of the games. This is just something pseudo-random data with uniform like distribution.

Reply:
Ugh yes of course I'm not sure what I was smoking here.

You can just instantiate a Random() instance with a fixed seed here.